### ISSUE TITLE
frustrating, mnist csv load problem (python)

### ISSUE THREAD

**Issue by channerduan01** [dmlc/mxnet#1445]
I used simple mnist problem to test mxnet. (data from kaggle: https://www.kaggle.com/c/digit-recognizer)
The data source is csv, so I preprocessed data and used below code to load:

batch = 30000
shape = (784,)
train = mx.io.CSVIter(
    data_csv='data.csv',
    data_shape=shape,
    label_csv='data_label.csv',
    label_shape=(1,),
    batch_size=batch)

There were no error, but the error rate in training set is higher than 90% no matter the batch.
I am sure 'data.csv' and 'data_label.csv' are right because I got good result from them by other machine learning method.
The demo code in mxnet got good answer:

train = mx.io.MNISTIter(
    image = "mnist/train-images-idx3-ubyte",
    label = "mnist/train-labels-idx1-ubyte",
    input_shape = shape,
    batch_size  = batch,
    flat        = True,
    )

And, the difference is the MXDataIter.getindex(). I do not know what the meaning of this index array. For the result of my code, the index is [0,1,2,3,...], however, for the result of demo code from mxnet, the index is [0,0,0,0,...] 

Could any body help me to load the mnist data from csv? (Acctually, it is the standard form in kaggle, I think it will be good for freshman like me to understand mxnet)

**Comment by tqchen** [comment]
Is the data shuffled? By default the csv read data in as it is and can read more size than the memory size, and did not shuffle the data by default. The index is not used during training.

As an alternative, MXNet do accept numpy array as input and I guess that is the easiest way if the data fit into memory.

**Comment by piiswrong** [comment]
batch size looks too big. learning rate is probably to big/small.

**Comment by channerduan01** [comment]
Thanks a lot! I dramatically reduced the learning rate and got good result.

It is interesting.
      1, If I use the MNIST data that your demo indicated (http://webdocs.cs.ualberta.ca/~bx3/data/mnist.zip, Binary files), the learning rate can be the default 0.1
      2, If I use the MNIST data from kaggle (https://www.kaggle.com/c/digit-recognizer, Csv files), the learning rate should be 0.001(even 0.003 was failed... with both batch sizes are 128)

But when I viewed the data from MXDataIter structure and even visualised them. Apart from index, they are similar.


Anyway, thank you! I love this elegant io system that makes the data loading much much easier!!!
